pkg/controller/applications/controller.go

```diff
@@ -124,21 +124,29 @@ func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.Ex
 		Name: &name,
 	}
 
+	if cr.Spec.ForProvider.Project != "" {
+		appQuery.Projects = []string{cr.Spec.ForProvider.Project}
+	}
+	if cr.Spec.ForProvider.AppNamespace != "" {
+		appQuery.AppNamespace = &cr.Spec.ForProvider.AppNamespace
+	}
+
 	// we have to use List() because Get() returns permission error
 	var apps *argocdv1alpha1.ApplicationList
 	apps, err := e.client.List(ctx, &appQuery)
 	if err != nil {
 		return managed.ExternalObservation{}, errors.Wrap(err, errListFailed)
 	}
-	app := &argocdv1alpha1.Application{}
-	for _, item := range apps.Items {
-		if item.Name == name && item.Spec.Project == cr.Spec.ForProvider.Project {
-			app = item.DeepCopy()
-		}
-	}
-	if app.Name == "" {
-		return managed.ExternalObservation{}, nil
+
+	if len(apps.Items) == 0 {
+		return managed.ExternalObservation{
+			ResourceExists:   false,
+			ResourceUpToDate: false,
+		}, nil
+	} else if len(apps.Items) > 1 {
```

Comment:
Don't need else here due to the return. Makes it easier to read.

```suggestion
	}
	if len(apps.Items) > 1 {
```

Reply:
Fixed